daffodil-core/src/main/scala/org/apache/daffodil/dsom/walker/AbstractDSOMWalker.scala

```diff
@@ -0,0 +1,174 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.daffodil.dsom.walker
+
+/**
+ * Below is a series of "View" Mixins defined such that particular simple attributes
+ * of the core DSOM classes are exposed as a Walk is performed.
+ *
+ * Several of them have a walkDSOM method.  By default, this method is called as the walk is performed,
+ * but a user implementing AbstractDSOMWalker can override this behavior by overriding walkFromRoot
+ * and implementing their own event handlers for each of the DSOM types.
+ */
+
+trait CommonContextView {
+  def namespaces: scala.xml.NamespaceBinding
+}
+
+trait TermView extends CommonContextView {
+  def isArray: Boolean
+  def isOptional: Boolean
+}
+
+trait TypeView
+trait SimpleTypeView extends TypeView {
+  def primType: PrimTypeView
+}
+trait ComplexTypeView extends TypeView {
+  def group: ModelGroupView
+}
+
+trait ModelGroupView extends TermView {
+  def groupMembers: Seq[TermView]
+}
+
+trait ChoiceView extends ModelGroupView
+trait SequenceView extends ModelGroupView
+trait GroupRefView extends ModelGroupView {
+  def isHidden: Boolean
```

Comment:
One last look over this (looking really good, btw), I'm wondering if we want isHidden visible? We could in theory have daffodil-core present a view of groupMembers where hidden group refs are filtered out. And then API users never even have to be aware that some things are hidden because they'll never see them. Right now that logic is handled by the AbstractDSOMWalker, but if a user didn't use that they'd have to keep hidden group refs in consideration.

I think the main issue with this is that daffodil-core can't actually modify the underlying groupMembers. Daffodil still DSOM still needs to know about hidden groups, and there is no way to actually present a View different from the concrete class. The View can only *hide* members--it can't actually make changes to those members.

I guess we just don't have an issue anywhere else where a View member needs a different type/content than the underlying member.

Not sure of a good way to handle this unless we start introduce different names for View members or allocating new classes, which probably starts getting messy. So maybe this isn't worth considering, and just something to keep in mind with how this View API works.

Reply:
I think there is a solution to this, but it would break the current paradigm of every field in the `View`s referencing an existing field.  We could replace the existing `group` element in `ComplexTypeView` with something like `optVisibleGroup`, and then the implementation for `optVisibleGroup` would pattern-match `group`; if it is hidden, return `None, otherwise return the `group` wrapped in an option.  Similarly, instead of `groupMembers` being in `TermView`, we would have `visisbleGroupMembers`, and in the implementation `visibleGroupMembers` would be a filtered list of the real `groupMembers`, excluding any `TermView` that is a `GroupRef` and is also hidden.

Now, my main concern with this is it would potentially cause more confusion, leading the end user to wonder why we're only allowed to see "visible" members, and might make traversal very annoying with having to constantly unpack Options.

Reply:
Yeah, I agree yoru concern, that does feel like things could get more confusing, and DSOM is alrady conufsing enough as it is.

An another approach with maybe less confusion would be to create case classes for all our Views. Accessing a View would result in allocating a new case class and copying relevant information from the base class into the View class, which would allow modifcations/filtering to what gets copied. Though, this could result in lots of allocations, which is probably best to avoided.

I'd say these types of changes probably aren't worth it unless we find more examples where it would be useful to have the View be substantially different. Probalby not worth it for just this one case.